merkle/compact/tree.go

```diff
@@ -71,55 +71,59 @@ type GetNodeFunc func(depth int, index int64) ([]byte, error)
 //
 // TODO(pavelkalinnikov): Make GetNodeFunc get all nodes at once.
 func NewTreeWithState(hasher hashers.LogHasher, size int64, f GetNodeFunc, expectedRoot []byte) (*Tree, error) {
-	sizeBits := bits.Len64(uint64(size))
-
-	r := Tree{
-		hasher: hasher,
-		nodes:  make([][]byte, sizeBits),
-		root:   hasher.EmptyRoot(),
-		size:   size,
+	ids := make([]nodeID, 0, bits.OnesCount64(uint64(size)))
+	for sz := uint64(size); sz != 0; sz &= sz - 1 { // Iterate 1-bits of size.
+		level := uint(bits.TrailingZeros64(sz))
+		index := (sz - 1) >> level
+		ids = append(ids, nodeID{level: level, index: index})
 	}
 
-	if isPerfectTree(size) {
+	hashes := make([][]byte, len(ids))
+	if len(ids) == 1 {
 		glog.V(1).Info("Is perfect tree.")
-		r.root = append(make([]byte, 0, len(expectedRoot)), expectedRoot...)
-		r.nodes[sizeBits-1] = r.root
+		hashes[0] = expectedRoot
 	} else {
-		// Pull in the nodes we need to repopulate our compact tree and verify the root
-		for depth := 0; depth < sizeBits; depth++ {
-			if size&1 == 1 {
-				index := size - 1
-				glog.V(1).Infof("fetching d: %d i: %d, leaving size %d", depth, index, size)
-				h, err := f(depth, index)
-				if err != nil {
-					glog.Warningf("Failed to fetch node depth %d index %d: %s", depth, index, err)
-					return nil, err
-				}
-				r.nodes[depth] = h
+		for i, id := range ids {
+			glog.V(1).Infof("Fetching node level %d index %d", id.level, id.index)
+			h, err := f(int(id.level), int64(id.index))
+			if err != nil {
+				glog.Warningf("Failed to fetch node level %d index %d: %v", id.level, id.index, err)
+				return nil, err
 			}
-			size >>= 1
+			// Note: Right border nodes of compact.Range are ordered top-to-bottom.
+			hashes[len(ids)-i-1] = h
 		}
-		r.recalculateRoot(func(depth int, index int64, hash []byte) error {
-			return nil
-		})
 	}
-	if !bytes.Equal(r.root, expectedRoot) {
-		glog.Warningf("Corrupt state, expected root %s, got %s", hex.EncodeToString(expectedRoot[:]), hex.EncodeToString(r.root[:]))
-		return nil, RootHashMismatchError{ActualHash: r.root, ExpectedHash: expectedRoot}
+
+	fact := RangeFactory{Hash: hasher.HashChildren}
+	rng, err := fact.NewRange(0, uint64(size), hashes)
+	if err != nil {
+		return nil, err
 	}
-	glog.V(1).Infof("Resuming at size %d, with root: %s", r.size, base64.StdEncoding.EncodeToString(r.root[:]))
-	return &r, nil
+
+	root, err := rng.GetRootHash()
+	if err != nil {
+		return nil, err
+	}
+	if size == 0 {
+		root = hasher.EmptyRoot()
+	}
+	t := Tree{hasher: hasher, rng: rng, root: root}
+
+	if !bytes.Equal(root, expectedRoot) {
+		glog.Warningf("Corrupt state, expected root %s, got %s", hex.EncodeToString(expectedRoot), hex.EncodeToString(root))
+		return nil, RootHashMismatchError{ActualHash: root, ExpectedHash: expectedRoot}
+	}
+
+	glog.V(1).Infof("Resuming at size %d, with root: %s", rng.End(), base64.StdEncoding.EncodeToString(root))
```

Comment:
This looks like a sequencer specific log message. Maybe something about restoring or loading the tree?